Long-Running Development Pull Request

DRMacIver opened the pull request:
As per the title: The development branch will eventually get merged into master when I am happy that it is nearing a releasable state. Up until then, this is the place to review any and all code, and generally question my ~life~ design choices!

DRMacIver committed:
Apparently I should not use should in specs

DRMacIver committed:
Add find for debug and shrinking specs

DRMacIver committed:
Workaround for Helix builds silently failing

DRMacIver committed:
Honesty in naming

DRMacIver committed:
Better rubocop rules

DRMacIver committed:
Implementation of new control loop + basic shrinking

jjgreen commented on spec/basic_examples_spec.rb:
```diff
@@ -4,16 +4,16 @@ def expect_failure(&block)
-RSpec.describe 'some basic hypothesis tests' do
-  it 'should think integer addition is commutative' do
+RSpec.describe 'basic hypothesis tests' do
+  it 'think integer addition is commutative' do
```
`it think`: your bad grammar tolerance is greater than mine, :smile:, of course I see the issue, `it` is intended for singular subjects, but here the subject is plural.  Rather than being forced to choose between grammatically correct stdout and code, why not alias `it` to `they`?

DRMacIver replied:
I like it. Will do.

DRMacIver committed:
Add code for calculating stable identifiers

DRMacIver committed:
Add alias for more grammatical tests

DRMacIver committed:
Start to modularise the rust code a bit

hwayne commented:
Or just alias it to "test" :P

On Sat, Feb 3, 2018 at 8:26 AM, David R. MacIver <notifications@github.com>
wrote:

> *@DRMacIver* commented on this pull request.
> ------------------------------
>
> In spec/basic_examples_spec.rb
> <https://github.com/HypothesisWorks/hypothesis-ruby/pull/12#discussion_r165815955>
> :
>
> > @@ -4,16 +4,16 @@ def expect_failure(&block)
>    expect(&block).to raise_exception(RSpec::Expectations::ExpectationNotMetError)
>  end
>
> -RSpec.describe 'some basic hypothesis tests' do
> -  it 'should think integer addition is commutative' do
> +RSpec.describe 'basic hypothesis tests' do
> +  it 'think integer addition is commutative' do
>
> I like it. Will do.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/HypothesisWorks/hypothesis-ruby/pull/12#discussion_r165815955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACiXdIGozOMX1FBtE2snJxaLpLYOMDswks5tRGyogaJpZM4R4JBQ>
> .
>

DRMacIver commented:
> Or just alias it to "test" :P

BUT HILLEL MY NATURAL LANGUAGE DSL!!!

DRMacIver committed:
Use a Result type internally even if we have to pass an option to Ruby

DRMacIver committed:
Improve Helix usage, implement providers in Rust

DRMacIver committed:
Add implementation of list provider

DRMacIver committed:
Track engine state for ease of debugging

DRMacIver committed:
Add minitest to dependencies

DRMacIver committed:
Reorder gems

DRMacIver committed:
Recover Example Printing

Somehow the merge with this code got lost.
This recovers it from git history.

DRMacIver committed:
Remove accidental duplication of given in merge

DRMacIver committed:
Don't separate local given and given, just track depth

DRMacIver committed:
Test for better behaviour of composite and printing

DRMacIver committed:
Be a responsible adult human being who resists the siren call of instance_eval

DRMacIver committed:
Rubocop, your opinions are bad

DRMacIver committed:
Whatever, Rubocop. Whatever

DRMacIver committed:
Move AWFUL-HACKS into docs

DRMacIver committed:
Add an actual design doc

DRMacIver committed:
Add some basic usage documentation

DRMacIver committed:
Move some things to possible omissions

DRMacIver committed:
Merge pull request #15 from HypothesisWorks/DRMacIver/providers

Start to sketch out how we can properly implement providers

DRMacIver committed:
Merge pull request #16 from HypothesisWorks/DRMacIver/list-provider

Add implementation of list provider

DRMacIver committed:
Merge pull request #18 from HypothesisWorks/DRMacIver/docs

Add some documentation to the repo

DRMacIver committed:
Always update state at beginning of function

DRMacIver committed:
Treat unexpected loop termination as a bug

DRMacIver committed:
Require 100% coverage and add supporting tests

DRMacIver committed:
Add less basic implementation of integers

DRMacIver committed:
Factor out Shrinker and loop shrinks to a fixed point

DRMacIver committed:
Merge pull request #17 from HypothesisWorks/DRMacIver/track-state

Track engine state for ease of debugging

DRMacIver committed:
Add implementations of integer bounds

DRMacIver committed:
Add mixed and choice_of providers

DRMacIver committed:
Merge pull request #19 from HypothesisWorks/DRMacIver/termination-is-error

Turn unexpected termination of main loop thread into an error

DRMacIver committed:
Merge pull request #20 from HypothesisWorks/DRMacIver/integers

Add less basic implementation of integers

DRMacIver committed:
Merge pull request #21 from HypothesisWorks/DRMacIver/better-shrinking

Factor out Shrinker and loop shrinks to a fixed point

DRMacIver committed:
I don't agree with that rule either

DRMacIver committed:
Merge branch 'DRMacIver/unions' into development

(This was meant to be https://github.com/HypothesisWorks/hypothesis-ruby/pull/22
but I screwed up with the base).

DRMacIver committed:
Implement debug and clone for the empty struct

DRMacIver committed:
Don't forget to update the count there!

DRMacIver committed:
Always read the attempt back from the result in case it got shrunk due to bit sizes

DRMacIver committed:
Ensure incorporate is only called correctly, add most basic of skipping logic

DRMacIver committed:
Give the main loop thread a name for better debugging

DRMacIver committed:
Don't provide an average_size parameter

DRMacIver committed:
Remove bits provider as no longer needed

DRMacIver committed:
Add some basic composition methods to Provider

DRMacIver committed:
Add a strings provider

DRMacIver committed:
Merge pull request #23 from HypothesisWorks/DRMacIver/strings

Add an actual working strings provider

DRMacIver commented:
Heads up @HypothesisWorks/hypothesis-ruby-contributors - I'm still going to keep doing pull requests into here because I like the workflow, but I don't realistically expect people to review them before they make it into here.

I'm hoping to get an alpha release end of next week - my plan is to get the providers library finished off today or tomorrow, at which point it will mostly come down to just figuring out the packaging. This may be anywhere between trivial (unlikely) to a major problem (hopefully not), but assuming it's somewhere closer to the former I think doing a release towards the end of the week should be viable.

The nice thing about alpha releases is that nothing is irrevocable! None of this API is committed to, and we can change whatever we like, yay! But it would still be nice not to release stuff and then decide that that was a terrible idea. So if you *do* want to look over this PR over the next week, that would be very much appreciated.

DRMacIver committed:
Fix formatting

DRMacIver committed:
They're called arrays, not lists, in ruby

DRMacIver committed:
patch HypothesisCoreRepeatValues to raise on a bad draw

DRMacIver committed:
Check for not using the argument to composite

DRMacIver committed:
Those are now unused

DRMacIver committed:
Add hashes

DRMacIver committed:
Also add fixed arrays

DRMacIver committed:
Merge pull request #25 from HypothesisWorks/DRMacIver/hashes

Add providers for hashes

DRMacIver committed:
Make debug helpers test only

DRMacIver committed:
No wonder I've been merging broken builds

DRMacIver commented:
Ha ha. So. Umm. Funny story @alexwlchan... 8d6375a26a6fe12b42ac36587fc93784225fcfaf

DRMacIver committed:
Add booleans provider

DRMacIver committed:
Merge pull request #27 from HypothesisWorks/DRMacIver/booleans

Add booleans provider

DRMacIver committed:
Merge pull request #26 from HypothesisWorks/DRMacIver/no-debug

Make debug helpers test only

DRMacIver committed:
Run tests from the built gem as the gods intended

DRMacIver committed:
It turns out that that wasn't actually working.

Add an extconf that specifies how to build the gem and errors if
cargo isn't installed.

DRMacIver committed:
I will not only suppress exceptions, I will suppress the rule that says I shouldn't

DRMacIver committed:
Don't require bundler in our Rakefile

DRMacIver committed:
Oops spaces not tabs

DRMacIver committed:
Merge pull request #28 from HypothesisWorks/DRMacIver/package-building

Build packages and run our tests against them

DRMacIver commented:
If you want to try this out, adding the following to your Gemfile should work:

```ruby
gem 'hypothesis-specs', :git => 'https://github.com/HypothesisWorks/hypothesis-ruby.git', :branch => 'development'
```

You will need a rust toolchain installed first! See https://www.rustup.rs/ for instructions.

jeremy-w commented on docs/design.md:
```diff
@@ -0,0 +1,118 @@
+RSPec.describe "integer addition" do
+  it "commutes" do
+    hypothesis do
+      m = given integers
```
The singular/plural mismatch here throws me a bit. I'd kinda expect either:

```ruby
ms = given integers
```

or something that makes more sense with a singular `m`, like:

```ruby
m = from integers
```

or

```ruby
m = given integer
```

DRMacIver replied:
Ugh. The pluralisation problem is *the worst*. The problem is that there are two contexts in which strategies can be used. `given an_integer` looks better than `given integers`, but `lists an_integer` looks much much worse than `lists integers`.

I'm kinda tempted to give each provider method a plural and a singular alias and make them completely interchangeable. Would that be awful?

hwayne replied:
If you go this way I think you should make the plural alias not have any articles, so `integer` and `integers`. Then you could make `a` and `an` noop messages to `given`, leaving you with `given an integer`.

DRMacIver replied:
Do you mean the singular alias?

As per comment below I've gone with `any_integer`. `an/a` doesn't really read right but...

Hmm...

How about `any` as a replacement name for `given`? It reads a lot better I think...

```ruby
hypothesis do
    x = any integer
    ls = any array(of: integers)
end
```

God that really is a lot better isn't it?

jeremy-w commented on docs/usage.md:
```diff
@@ -0,0 +1,175 @@
+* A set of "assumptions" - things that need to be true in
+  order for the test to be valid.
+
+The givens come from *providers*, which describe a range of
```
Vaguely wondering if "givers" or "sources" might be clearer as a name than "providers".

DRMacIver replied:
All of the possible names for this are terrible. :-( [Some prior art on this discussion](https://github.com/HypothesisWorks/hypothesis-python/issues/804)

For the specific suggestions: "givers" sounds hella awkward, and the probelm with "sources" is that a) It's already used elsewhere in the API (though internally and b) They're not *really* a source of example, they're a description of how an example could be built and the source of the example is the test case itself.

jeremy-w commented on docs/usage.md:
```diff
@@ -0,0 +1,175 @@
+## Defining your own providers
+
+The library of providers you get from Hypothesis is large but finite.
+At some point you're going to want something it doesn't include - e.g. to
```
This reads like it's kinda unusual to build your own provider. Is it expected to generate-and-prune with assume more often than just build a provider that doesn't need further pruning most of the time?

DRMacIver replied:
No, that reading is unintentional. Will try to fix, thanks.

It's super-normal to build your own providers, there's just also a large(ish) built in library for doing so.

jeremy-w commented on docs/usage.md:
```diff
@@ -0,0 +1,175 @@
+RSpec.describe User do
+  it "has a name" do
+    hypothesis do
+      expect(given(user).name.length).to be > 0
```
Should this be `given(users)` to match the plural of `def users` up at L130?

(And doesn't it really seem like `given(singular_noun)` is more natural?)

jeremy-w replied:
This reminds me that I often name test stub data as `any_xyz`, like `given(any_user)` to clarify that it's not a specific user. Perhaps that prefixed `any` would be a clear key that it's a provider and may vary across that domain?

DRMacIver replied:
Hmm. `any_` as a prefix for providers isn't a bad idea at all. Lets see...

```ruby
given(any_integer)
given(any_list(any_integer))
given(any_fixed_dictionary(a: any_boolean))
```

Hmmm. I think I like it!

DRMacIver committed:
Very basic yard setup

DRMacIver committed:
Add missing LICENSE file

DRMacIver committed:
Many documentation

DRMacIver committed:
Update README

DRMacIver committed:
That is now redundant

DRMacIver committed:
Remove moved class

DRMacIver committed:
Fix include

DRMacIver committed:
The naming convention is dead, long live the naming convention

DRMacIver committed:
Formatting

DRMacIver committed:
Merge pull request #29 from HypothesisWorks/DRMacIver/documentation

Improve documentation and naming conventions

DRMacIver committed:
Add task for gem building

DRMacIver committed:
Set up release uploading

DRMacIver committed:
Add rubygems deploy provider

DRMacIver committed:
Formatting

DRMacIver committed:
Merge pull request #30 from HypothesisWorks/DRMacIver/gem-building

Building and releasing gems

DRMacIver commented:
Notable update: providers now have a weird dual name thing going on where there is both `any_foo` and `foos` for most provider names.

I'm not totally happy with this decision but I think it might be least-worst.

DRMacIver committed:
Update design doc a bit

DRMacIver committed:
More API golfing for readability

DRMacIver committed:
Rename mixed with better aliases

DRMacIver committed:
Merge pull request #32 from HypothesisWorks/DRMacIver/any

More API golfing for readability

DRMacIver committed:
Better name

DRMacIver committed:
The API really reads much better this way, though the control flow is now quite weird

DRMacIver committed:
Everything looks so much nicer with global variables!

DRMacIver committed:
Support passing block via any

DRMacIver committed:
Better name than composite

DRMacIver committed:
Switch over from Provider to Possible

DRMacIver committed:
Merge pull request #33 from HypothesisWorks/DRMacIver/api-polishing

A bunch of polishing of the API

alexwlchan commented on lib/hypothesis.rb:
```diff
@@ -2,29 +2,217 @@
+  # @param name [String, nil] An optional name to show next to the result on
+  #   failure. This can be helpful if you have a lot of givens in your
+  #   hypothesis, as it makes it easier to keep track of which is which.
+  def any(provider, name: nil, &block)
```
Just on an idle skim through this PR, it seems weird that a parameter called `provider` takes an instance of `Possible` – and there are a bunch of other lines where you still use the old terminology.

DRMacIver replied:
Whoops. Thought I caught them all, but obviously not! Will do another pass through and grep properly this time, thanks.

DRMacIver committed:
Fix remaining references to provider

DRMacIver committed:
Merge pull request #34 from HypothesisWorks/DRMacIver/correct-vestigial-providers

Fix remaining references to provider

DRMacIver commented:
I'm inclined to revise my earlier timeline and soft-release this tomorrow in its capacity as very early days developer preview. The timeline was based on the expectation of packaging taking ages, and in the I've got further this weekend than I was planning to by the end of the week!

This doesn't imply that this API is it, there's still plenty of scope for change, but at this point I think it's hard for people to give useful feedback without using it, and having it pushed to rubygems.org will make that easier and will mean it's e.g. available on rubydoc.info

suzannehamilton commented on README.markdown:
```diff
@@ -1,5 +1,82 @@
+
+      to_remove = any element_of(values)
+
+      values.delete_at(value.index(to_remove))
```
Typo: `value.index` should be `values.index`.

suzannehamilton commented on lib/hypothesis/possible.rb:
```diff
@@ -0,0 +1,369 @@
+    # shape.
+    # This is used for hashes where you know exactly what the
+    # keys are, and different keys may have different possible values.
+    # For example, fixed_hashes(a: integers, b: booleans)
```
Possibly a stale comment? Was `fixed_hashes` the old name for this function?

Ditto for `fixed_arrays` below.

DRMacIver replied:
Yup, those were name changes where I failed to update the comment. Thanks, will fix.

suzannehamilton reviewed (commented):
I had a quick play with the current implementation out of curiosity, and I think the API flows very nicely. (With the caveat that I'm not enough of a Rubyist to judge whether it's idiomatic. I'm still slowly coming round to rspec.)

I spotted a couple of typos and old references in the docs - see the other comments.

By the way, I'm assuming that it's expected behaviour for the shrinker to repeatedly try out the same values before it finds a good example. Is that just one of the ways in which the shrinker is currently very primitive? If that _does_ sound surprising, though, let me know and I'll give an example.

DRMacIver commented:
Thanks for the feedback, Suzy!

>  I'm not enough of a Rubyist to judge whether it's idiomatic. I'm still slowly coming round to rspec

Yeah, same really (I never really came around to RSpec despite about 4 years of prior use, but it's improved a lot in my five year absence)

> By the way, I'm assuming that it's expected behaviour for the shrinker to repeatedly try out the same values before it finds a good example. Is that just one of the ways in which the shrinker is currently very primitive? If that does sound surprising, though, let me know and I'll give an example.

This is mostly expected behaviour. It's also mostly expected behaviour that it would worse than normal because of the current rather basic shrinker.

Basically there are two things going on:

1. The same test case can have multiple different underlying representations and the shrinker will try to move to a simpler representation even if that doesn't change the givens of the test case.
2. The shrinker is *supposed* to avoid executing duplicate underlying representations, but it currently doesn't, so will often end up trying the same attempted shrink that it's previously tried and determined not to be a failing test case.

DRMacIver committed:
Fix typo

DRMacIver committed:
Fix references to old method names

DRMacIver commented:
:shipit:
